pkg/kapis/events/v1alpha2/handler.go

```diff
@@ -0,0 +1,191 @@
+		}
+		ar.Events, err = h.eo.SearchEvents(filter, from, size, sorts)
+	}
+
+	if err != nil {
+		klog.Errorln(err)
+		api.HandleInternalError(resp, req, err)
+		return
+	}
+	resp.WriteAsJson(ar)
+}
+
+func (h handler) userScope(user user.Info) (workspaces []*v1alpha1.Workspace, namespaces []*corev1.Namespace, err error) {
+	lw, e := h.tenant.ListWorkspaces(user, query.New())
+	if e != nil {
+		err = fmt.Errorf("error listing workspaces: %s", e)
+		return
+	}
+	for _, r := range lw.Items {
+		if ws, ok := r.(*v1alpha1.Workspace); ok {
+			workspaces = append(workspaces, ws)
+		}
+	}
+	ln, e := h.tenant.ListNamespaces(user, "", query.New())
+	if e != nil {
+		err = fmt.Errorf("error listing namespaces: %s", e)
+		return
+	}
+	for _, r := range ln.Items {
+		if ns, ok := r.(*corev1.Namespace); ok {
+			namespaces = append(namespaces, ns)
+		}
+	}
+	return
+}
+
+func (h handler) parseFilterFrom(req *restful.Request, userWorkspaces []*v1alpha1.Workspace,
+	userNamespaces []*corev1.Namespace) (*events.Filter, error) {
+	queryNamespaces := stringutils.Split(req.QueryParameter("involved_object_namespace_filter"), ",")
+	queryNamespaceSubstrs := stringutils.Split(req.QueryParameter("involved_object_namespace_search"), ",")
+	queryWorkspaces := stringutils.Split(req.QueryParameter("workspace_filter"), ",")
+	queryWorkspaceSubstrs := stringutils.Split(req.QueryParameter("workspace_search"), ",")
+	nsMap, err := h.intersectedNamespaces(userWorkspaces, userNamespaces,
+		queryWorkspaces, queryNamespaces, queryWorkspaceSubstrs, queryNamespaceSubstrs)
+	if err != nil {
+		return nil, err
+	}
+
+	filter := &events.Filter{InvolvedObjectNamespaceMap: nsMap}
+	filter.InvolvedObjectNames = stringutils.Split(req.QueryParameter("involved_object_name_filter"), ",")
+	filter.InvolvedObjectNameFuzzy = stringutils.Split(req.QueryParameter("involved_object_name_search"), ",")
+	filter.InvolvedObjectkinds = stringutils.Split(req.QueryParameter("involved_object_kind_filter"), ",")
+	filter.MessageFuzzy = stringutils.Split(req.QueryParameter("message_search"), ",")
+	filter.Type = req.QueryParameter("type_filter")
+	filter.Reasons = stringutils.Split(req.QueryParameter("reason_filter"), ",")
+	filter.ReasonFuzzy = stringutils.Split(req.QueryParameter("reason_search"), ",")
+
+	if tstr := req.QueryParameter("start_time"); tstr != "" {
+		t, err := time.Parse(time.RFC3339, tstr)
+		if err != nil {
+			return nil, err
+		}
+		filter.StartTime = &t
+	}
+	if tstr := req.QueryParameter("end_time"); tstr != "" {
+		t, err := time.Parse(time.RFC3339, tstr)
+		if err != nil {
+			return nil, err
+		}
+		filter.EndTime = &t
+	}
+	return filter, nil
+}
+
+func (h handler) intersectedNamespaces(userWorkspaces []*v1alpha1.Workspace, userNamespaces []*corev1.Namespace,
+	queryWorkspaces, queryNamespaces, queryWorkspaceSubstrs, queryNamespaceSubstrs []string) (map[string]time.Time, error) {
+	var workspaces []string
+	for _, ws := range userWorkspaces {
```

Comment:
Maybe we can turn queryNamespaces/queryWorkspaces to a  map and then do the search.

a slice search is too slow